product_pricelist_supplierinfo/models/product_pricelist.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2018 Tecnativa - Vicent Cubells
+# Copyright 2018 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import fields, models
+
+
+class ProductPricelist(models.Model):
+    _inherit = "product.pricelist"
+
+    def _compute_price_rule(self, products_qty_partner, date=False, uom_id=False):
+        """Recompute price after calling the atomic super method for
+        getting proper prices when based on supplier info.
+        """
+        rule_obj = self.env["product.pricelist.item"]
+        result = super()._compute_price_rule(products_qty_partner, date, uom_id)
+        # Make sure all rule records are fetched at once at put in cache
+        rule_obj.browse(x[1] for x in result.values()).mapped("price_discount")
+        for product, qty, _partner in products_qty_partner:
+            rule = rule_obj.browse(result[product.id][1])
+            if rule.compute_price == "formula" and rule.base == "supplierinfo":
+                context = self.env.context
```

Comment:
@victoralmau You could put context assignation as it would not change between two iterations of the loop